pilot/pkg/bootstrap/server.go

```diff
@@ -389,8 +390,23 @@ func (s *Server) WaitUntilCompletion() {
 }
 
 // initKubeClient creates the k8s client if running in an k8s environment.
+// This is determined by the presence of a kube registry, which
+// uses in-context k8s, or a config source of type k8s.
 func (s *Server) initKubeClient(args *PilotArgs) error {
-	if hasKubeRegistry(args.RegistryOptions.Registries) {
+	hasK8SConfigStore:= false
+	if args.RegistryOptions.FileDir == "" {
+		// If file dir is set - config controller will just use file.
+		meshConfig := s.environment.Mesh()
+		if meshConfig != nil && len(meshConfig.ConfigSources) > 0 {
+			for _, cs := range meshConfig.ConfigSources {
+				if cs.Address == "k8s://" {
+					hasK8SConfigStore = true
```

Comment:
nit: this seems to make more sense to be part of the `hasKubeRegistry` function? Any reason to prefer here

Reply:
It's not technically checking if it has kube registry - this is a kube config source.

Maybe rename hasKubeRegistry to 'needsK8S' ? And also check if CA is enabled.